A lot of changes: Kotlin, time/date formatting fixes, demo activity for whole view snapping...

AndroidDeveloperLB opened the pull request:
I've prepared a pull request that allows to snap by entire view (week by week, for example) based on [this pull request](https://github.com/Quivr/Android-Week-View/pull/88), together with some fixes of issues I've found, and conversion to 100% Kotlin code.

Sadly, some issues that I've found still exist:
- https://github.com/Quivr/Android-Week-View/issues/31#issuecomment-387005542 - slow on the normal sample activities. No idea why they occur (maybe the dragging feature?), but the new activity doesn't have those.
- https://github.com/alamkanak/Android-Week-View/issues/495 - the RTL direction issue.
- https://github.com/alamkanak/Android-Week-View/issues/494 - shortened format even though there is enough space
-https://github.com/Quivr/Android-Week-View/issues/31#issuecomment-387005542 - snapping can be to part of the day.

My repository is here:
https://github.com/AndroidDeveloperLB/Android-Week-View
You can use latest commit in the project, from here:
https://jitpack.io/#AndroidDeveloperLB/Android-Week-View

AndroidDeveloperLB committed:
-updated sdks
-added ability to scroll whole view (entire week, for example) like a ViewPager.
-made default first day of week to be the one of the calendar.

AndroidDeveloperLB committed:
-converted project to Kotlin. Code can be shortened a lot, but requires some delicate work.
-fixed some time/date formatting related issues:
https://github.com/alamkanak/Android-Week-View/issues/497
https://github.com/alamkanak/Android-Week-View/issues/495 (but not fixed RTL alignment issue)
-created a new activity to demonstrate the paging of entire view (example: week by week snapping), based on this pull request:
https://github.com/Quivr/Android-Week-View/pull/88

SkyleKayma commented on sample/build.gradle:
```diff
@@ -22,6 +21,6 @@ android {
     compile project(':library')
-    compile 'com.android.support:appcompat-v7:25.1.0'
+    compile 'com.android.support:appcompat-v7:27.1.1'
     compile 'com.squareup.retrofit:retrofit:1.9.0'
```
Pls go with new version since everything has been updated, why not that?
`implementation 'com.squareup.retrofit2:retrofit:2.4.0'`
And change all "compile" by "implementation".

AndroidDeveloperLB replied:
Makes sense. Added.
The retrofit one is problematic. API has changed, and I don't want to see what should be changed just for this. Need to do other work... Will stay with 1.9.0 for now.

SkyleKayma replied:
Ok ok ;D

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/MonthLoader.kt:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    override fun onLoad(periodIndex: Int): List<WeekViewEvent>? {
+        return onMonthChangeListener!!.onMonthChange(periodIndex / 12, periodIndex % 12 + 1)
```
How can you be sure that "onMonthChangeListener" will not be null here ?
I think that nullable "?" is better. It returns nullable List btw.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekView.kt:
```diff
@@ -0,0 +1,2513 @@
+            invalidate()
+        }
+    private var mTimeTextPaint: Paint? = null
+    private var mTimeTextWidth: Float = 0.toFloat()
```
don't use this : 
`0.toFloat()`
Use this instead : 
`0f`
If you can do the change everywhere

AndroidDeveloperLB replied:
Right. This was part of the auto-conversion. Fixed that too.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekView.kt:
```diff
@@ -0,0 +1,2513 @@
+                        canvas.drawRect(start, startY, startPixel + mWidthPerDay, height.toFloat(), futurePaint!!)
+                    }
+                } else {
+                    canvas.drawRect(start, mHeaderHeight + (mHeaderRowPadding * 2).toFloat() + mTimeTextHeight / 2 + mHeaderMarginBottom, startPixel + mWidthPerDay, height.to
```
Try to avoid usage of reserved words. If you could rename this by something like "value", it could be great.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.kt:
```diff
@@ -0,0 +1,206 @@
+    constructor(id: String, name: String, startYear: Int, startMonth: Int, startDay: Int, startHour: Int, startMinute: Int, endYear: Int, endMonth: Int, endDay: Int, endHour: Int, endMinute: Int) {
+        this.identifier = id
+
+        this.startTime = Calendar.getInstance()
```
In Kotlin you have the "apply" function that allow you a better way to do: 
`this.startTime = Calendar.getInstance().apply{
    set(Calendar.YEAR, startYear)
    set(Calendar.MONTH, startMonth - 1)
    set(Calendar.DAY_OF_MONTH, startDay)
    set(Calendar.HOUR_OF_DAY, startHour)
    set(Calendar.MINUTE, startMinute)
}`

SkyleKayma replied:
Let a line jump between lines, can't do it on Github.

AndroidDeveloperLB replied:
Seems ok. Done

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekViewEvent.kt:
```diff
@@ -0,0 +1,206 @@
+        this.startTime!!.set(Calendar.HOUR_OF_DAY, startHour)
+        this.startTime!!.set(Calendar.MINUTE, startMinute)
+
+        this.endTime = Calendar.getInstance()
```
Same here, use "apply"

AndroidDeveloperLB replied:
Done

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekViewUtil.kt:
```diff
@@ -21,22 +21,24 @@
-        return today;
+    @JvmStatic
+    fun today(): Calendar {
+        val today = Calendar.getInstance()
```
One line code, return all the line and use "apply"

SkyleKayma commented on sample/src/main/AndroidManifest.xml:
```diff
@@ -1,28 +1,24 @@
-        android:label="@string/app_name"
-        android:theme="@style/AppTheme">
+        android:allowBackup="true" android:icon="@drawable/ic_launcher" android:label="@string/app_name"
+        android:theme="@style/AppTheme" tools:ignore="AllowBackup,GoogleAppIndexingWarning">
```
Do not use `tools:ignore="AllowBackup"`, it prevents the modification of this attribute in projects that will use this lib. Prefer `tools:replace="allowBackup"`

AndroidDeveloperLB replied:
I don't think this is correct. It's the sample module. Not the library. Are you sure about this?

SkyleKayma replied:
Yeah you are right.

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+        setContentView(R.layout.activity_base)
+
+        val draggableView = findViewById<View>(R.id.draggable_view) as TextView
+        draggableView.setOnLongClickListener(DragTapListener())
```
In kotlin, you just have to use like this : 
`draggable_view.setOnLongClickListener(DragTapListener())`

No need findViewById

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+
+
+        // Get a reference for the week view in the layout.
+        weekView = findViewById<View>(R.id.weekView) as WeekView
```
Same here, this line is not needed anymore.

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+
+        // Set minDate
+        /*Calendar minDate = Calendar.getInstance();
+        minDate.set(Calendar.DAY_OF_MONTH, 1);
```
Even if it's commented, we could probably use the "apply" function.

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+        mWeekView.setMinDate(minDate);
+
+        // Set maxDate
+        Calendar maxDate = Calendar.getInstance();
```
Same here, use "apply"

AndroidDeveloperLB replied:
It's a part of a comment... And it's in the sample... And, I don't think it helps much here

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+        maxDate.set(Calendar.DAY_OF_MONTH, 10);
+        mWeekView.setMaxDate(maxDate);
+
+        Calendar calendar = (Calendar) maxDate.clone();
```
Same here again, use "apply"

AndroidDeveloperLB replied:
It's a part of a comment... And it's in the sample... And, I don't think it helps much here

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BaseActivity.kt:
```diff
@@ -0,0 +1,252 @@
+     * @param shortDate True if the date values should be short.
+     */
+    private fun setupDateTimeInterpreter(shortDate: Boolean) {
+        val calendar = Calendar.getInstance()
```
Use "apply"

AndroidDeveloperLB replied:
Fine.

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/BasicActivity.kt:
```diff
@@ -0,0 +1,188 @@
+        // Populate the week view with some events.
+        val events = ArrayList<WeekViewEvent>()
+
+        var startTime = Calendar.getInstance()
```
Use "apply"

SkyleKayma replied:
And do the same for each lines after pls ;D

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/MainActivity.kt:
```diff
@@ -0,0 +1,32 @@
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        findViewById<View>(R.id.buttonBasic).setOnClickListener {
```
Don't use findViewById, use directly `buttonBasic.setOnClickListener`

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/MainActivity.kt:
```diff
@@ -0,0 +1,32 @@
+            startActivity(intent)
+        }
+
+        findViewById<View>(R.id.buttonAsynchronous).setOnClickListener {
```
Same here, don't use findViewById

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/MainActivity.kt:
```diff
@@ -0,0 +1,32 @@
+            val intent = Intent(this@MainActivity, AsynchronousActivity::class.java)
+            startActivity(intent)
+        }
+        findViewById<View>(R.id.buttonWholeViewSnap).setOnClickListener {
```
Same here, don't use findViewById

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/WholeViewSnappingActivity.kt:
```diff
@@ -0,0 +1,38 @@
+        weekView.setMaxTime(24)
+        weekView.isScrollNumberOfVisibleDays = true
+        setDayViewType(TYPE_WEEK_VIEW)
+        val cal = Calendar.getInstance()
```
Use "apply"

SkyleKayma commented on sample/src/main/java/com/alamkanak/weekview/sample/WholeViewSnappingActivity.kt:
```diff
@@ -0,0 +1,38 @@
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        if (item.itemId == R.id.action_today) {
+            val cal = Calendar.getInstance()
+            cal.set(Calendar.DAY_OF_WEEK, cal.firstDayOfWeek)
```
Use apply

AndroidDeveloperLB replied:
Fine. Though I think the original code is easier to read.

SkyleKayma commented:
I did not try your code actually. But there is some changes to do in the Kotlin code. I will try to test it this afternoon. I probably forgot to notify things in the code, so if someone has the courage to double check everything, that would be great.

SkyleKayma commented on sample/src/main/res/layout/activity_main.xml:
```diff
@@ -1,23 +1,17 @@
+<LinearLayout
+    xmlns:android="http://schemas.android.com/apk/res/android" xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent" android:layout_height="match_parent" android:background="#fff"
+    android:orientation="vertical" tools:context=".MainActivity">
```
If you could avoid to put all values in the same line, it could be great

SkyleKayma replied:
Same for after

AndroidDeveloperLB replied:
It's just the style we have in other projects. Can I set it to be the standard one just in this project? I don't want to set a global rule...

SkyleKayma replied:
Let's push like this, I could do this later ;)

AndroidDeveloperLB replied:
I actually prefer the way it is now, because it's easier to see view hierarchy even in XML. Many times, when a view has a lot of attributes, it causes me to scroll a lot and try to remember what's happening around. Hard to read and understand what's going on, while also seeing the attributes. 

But I do understand the request: it's the standard. In the past I also wanted to use WhiteSmith code style for Java, but I left it in the end because I couldn't convince anyone to use it (I still think it's better than all the rest).

AndroidDeveloperLB commented:
@SkyleKayma I've now tested this fork:
https://github.com/openium/Android-Week-View
It has the snapping of whole view (week by week snapping, for example), without the issue I've written about here : https://github.com/Quivr/Android-Week-View/issues/31#issuecomment-387005542
But it has a serious issue when I switch to 3-days. Maybe it's fixated on 7 days...
Can you please check it out? Maybe it can help here to solve this issue?

About Kotlin, I think it can be shortened a lot, but as the code became more complex, and because I've done it multiple times already with the other forks (tried various ones till I got here), I decided to do less this time.

SkyleKayma commented:
Can you explain to me what is the issue in 3days view with the changes you had with this PR ?
I tried the code given here, and all is working correctly.

You are on the real fork actually. It's the last one that is active. So if you need to do some changes with Kotlin code, it's clearly in this repo.

AndroidDeveloperLB commented:
@SkyleKayma My code doesn't have this issue. The 3-days view issue is of this repository that I've tested:
https://github.com/openium/Android-Week-View
The issue is that is snaps very bad in this case. I think it might try to snap every 7 days even if you are in this state.
I mentioned it because it doesn't have this issue that does exist in my code (and before I forked) : 
https://github.com/Quivr/Android-Week-View/issues/31#issuecomment-387005542
If you know how to fix it, please let me know.

About the fork, I've now worked on even more changes and fixes. 

One fix is of a bug I didn't even report yet, that the day-month formatting in the header doesn't consider the one of the device. In many countries, the day is before the month, so you should see something like "15/4" instead of "4/15" . 
In the current implementation, it's always month and then day.

I really hope I can fix the other issues too.

jhoobergs commented:
@AndroidDeveloperLB First, thanks for your PR
It's hard to review such a large PR in which java is changed to kotlin etc. 

I think I would merge this PR if you added tests (which we still don't have 😞 ) so we can be more certain about it functioning correct.

SkyleKayma commented:
I'm probably stupid (or I missed something), but why want to fix a bug that you do not have in your code? :/
This repo (https://github.com/openium/Android-Week-View) is in any case no longer maintained and has several merge that have never been added. (I will probably delete this one, if this PR is accepted)

Btw, your code is working well. If you could do the little changes I said before, it could be awesome. (Or tell me if you are not agree with what I wrote)

AndroidDeveloperLB commented:
I don't want to fix a bug that I don't have in the code. I want to fix a bug that I do have in the code (and in the original code, before I even forked it).
It's just that the bug exists in my code, but not in his, so I thought maybe you could check it out and get a clue of how to fix it.
The other repository has a different bug, so I mentioned it too.

This bug is quite serious compared to the RTL and shortened text issues and easy to reproduce, so if you can fix it, please let me know.

About my code working well, I did it in a delicate way, after having experienced with the same procedure on other forks here (we want all code of ours to be in Kotlin, so I wanted to do it here too).
There will be a new pull request tomorrow with fewer changes:
- The fix I've mentioned of day+month order
- minor code cleaning 
- adding 2 new design-related (optional) features : write on the left of the week-days, have only the week-days to have a background (meaning the all-day events will have normal background).

AndroidDeveloperLB commented:
@jhoobergs I never have time for unit tests (literally) . Not on my free time and not at the office. We are just too few developers who are in tight schedule.

AndroidDeveloperLB committed:
-added fix for day&month formatting, to be according to current locale:
https://github.com/Quivr/Android-Week-View/issues/98
-added a feature to put text next to the week days
-added a feature to set only the week days to have a background, instead of week-days and all-day events.
-replaced Java "List" class with "MutableList" instead of Kotlin "List", to avoid possible crashes at runtime (because Kotlin "List" cannot be modified and is more restricted). That being said, I think in most cases, it's ok to use Kotlin "List", because in many cases you don't really need to modify the lists.
-some code cleaning and refactoring, including conversion of some fields to be Kotlin properties (setters and getters near the field itself)

AndroidDeveloperLB commented:
OK here I've made a new commit with those changes.
Waiting for accepting the pull request :)

SkyleKayma commented:
Where are the changes about what I have notified before ? (findViewById, apply, retrofit 2, 0f instead of 0.toFloat(), tools:replace instead of tools:ignore ...)

AndroidDeveloperLB commented:
@SkyleKayma I don't know what you are talking about. Please explain.

AndroidDeveloperLB committed:
Fixed possible null issues with providing null as list of events.
replaced findViewById with synthetic ones of Kotlin.

AndroidDeveloperLB commented:
@SkyleKayma But speaking of findViewById, I decided to replace them all :)

SkyleKayma commented:
Take a look at the review I started to do above in this discussion ;D
I had some comments when you probably need to change something.

The findViewById was one of these things ;)

SkyleKayma commented:
Oh ! My bad, I totaly forgot to submit the review x) Now it's done !(Some comments are outdated now, because I wrote this Yesterday x) )

AndroidDeveloperLB committed:
Cleaned code a bit.
Removed old fonts, using Lato font, which is more modern and used on Android's latest versions.
fixed issues of fonts, which didn't apply to all texts.
getting fonts by using the support library.

AndroidDeveloperLB commented:
ok done with all the comments you've written, and also fixed some of my own.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/MonthLoader.kt:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    override fun onLoad(periodIndex: Int): List<WeekViewEvent>? {
+        return onMonthChangeListener!!.onMonthChange(periodIndex / 12, periodIndex % 12 + 1)
```
How can you be sure that "onMonthChangeListener" will not be null here ?
I think that nullable "?" is better. It returns nullable List btw.

AndroidDeveloperLB replied:
I think you are both correct and incorrect. For some reason the library has code to check it that when you use it, you must set a listener:

        if (weekViewLoader == null && !isInEditMode)
            throw IllegalStateException("You must provide a MonthChangeListener")

and then when you set it:

    var monthChangeListener: MonthLoader.MonthChangeListener?
        get() = if (weekViewLoader is MonthLoader) (weekViewLoader as MonthLoader).onMonthChangeListener else null
        set(monthChangeListener) {
            this.weekViewLoader = MonthLoader(monthChangeListener)
        }

So you can't escape it from being non null.

But why this requirement exists at all?
How odd...
I think it's better to remove the requirement, and check for null...

For now I will let it stay how it is.

SkyleKayma commented:
You did great dude. Could you take a look just above ?

SkyleKayma commented:
You did great dude. Could you take a look just above ?

AndroidDeveloperLB committed:
Handled some warnings...

AndroidDeveloperLB commented:
Done. Please, if you know how to fix the snapping issue, let me know. I've worked a lot on this library all around, but couldn't find why this occurs.
I also need to do some very special customizations that I don't think should be in the library. I've contributed what I can and think that should be here.

AndroidDeveloperLB committed:
further consideration of current locale to use the best format for the dates being shown in the headers.

AndroidDeveloperLB committed:
handled auto-refreshing of the view every minute, to reflect changes of current time&date. You can test it by setting the time of the day to be 23:59, and wait a minute...
cleaned code a bit.

AndroidDeveloperLB committed:
Some code cleaning

AndroidDeveloperLB committed:
Using proper conversion of dp units as the default values of the WeekView.

AndroidDeveloperLB commented:
@SkyleKayma Can you please confirm the pull request?
I know I keep adding more and more, but it's just because I find more and more things that could be better or are just wrong.

Currently I want to work on the special customization I'm required to do.

AndroidDeveloperLB committed:
Replaced "compile" with "implementation

AndroidDeveloperLB committed:
minor tweaks to sample of the WholeViewSnappingActivity.
added a way to draw the line seperators horizontally&vertically in a way that makes them as in "checkers" (rectangles), by using isUsingCheckersStyle.
This is demonstrated in WholeViewSnappingActivity.

AndroidDeveloperLB committed:
reduced Calendar instance creation a bit

SkyleKayma commented:
It seems ok. I don't have the right to merge this PR. For that you need to ask @jhoobergs .

For the issue you said before, you just need to do the changes I will tell you in the next review.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekView.kt:
```diff
@@ -0,0 +1,2525 @@
+    private val mGestureListener = object : GestureDetector.SimpleOnGestureListener() {
+
+        override fun onDown(e: MotionEvent): Boolean {
+            mStartOriginForScroll = mCurrentOrigin.x
```
Delete the line `mStartOriginForScroll = mCurrentOrigin.x` This don't need to be done at each onDown.

SkyleKayma commented on library/src/main/java/com/alamkanak/weekview/WeekView.kt:
```diff
@@ -0,0 +1,2525 @@
+                if (mShowDistinctPastFutureColor) {
+                    val isWeekend = day.get(Calendar.DAY_OF_WEEK) == Calendar.SATURDAY || day.get(Calendar.DAY_OF_WEEK) == Calendar.SUNDAY
+                    val pastPaint = if (isWeekend && mShowDistinctWeekendColor) mPastWeekendBackgroundPaint else mPastBackgroundPaint
+     
```
After mCurrentOrigin.x = ...
just add the line deleted just before : 
`mStartOriginForScroll = mCurrentOrigin.x`

AndroidDeveloperLB commented:
@SkyleKayma Can you ... do a pull request on my repository for this fix?

Commit:
Fix scroll snapping

SkyleKayma commented:
Done, you can accept and test it.

AndroidDeveloperLB committed:
Merge pull request #1 from SkyleKayma/ScrollFix

Fix scroll snapping

AndroidDeveloperLB commented:
@SkyleKayma Thank you so much!
It works great now.
Now only the RTL issues are left. I hope I will have time to fix those too.

Do you know perhaps why sometimes the content changes from long to short formats (reported here: https://github.com/Quivr/Android-Week-View/issues/92 ) ? And perhaps how to fix it?

AndroidDeveloperLB committed:
-a lot of conversions (and fixes) from fields to properties.
-renamed and fixed some fields that are related to padding/margins, to make it clear what they are intended for.

AndroidDeveloperLB committed:
-a lot of conversions (and fixes) from fields to properties.
-renamed and fixed some fields that are related to padding/margins, to make it clear what they are intended for.

AndroidDeveloperLB committed:
Merge branch 'develop' of https://github.com/AndroidDeveloperLB/Android-Week-View into develop

AndroidDeveloperLB committed:
sample bug fix (reported here: https://github.com/Quivr/Android-Week-View/issues/92 )
fixed issue of inconsistency of short vs long text.
Now will show short text whenever you switch to week view.

AndroidDeveloperLB committed:
Refactored some attributes to make it easier to understand what they mean.
added ability to set subtitle text below the week-day title text.
Demonstrated in WholeViewSnappingActivity.
replaced deprecated clipRect function with the more recommended one, as written here: https://stackoverflow.com/q/50231950/878126

AndroidDeveloperLB committed:
fixed text alignment and spacing issues

AndroidDeveloperLB commented:
The issue of the content changes from long to short formats is just in the sample. Anyway, I've recently added a way to have multiple lines on the week-day area.

AndroidDeveloperLB committed:
-removed dropping from WholeViewSnappingActivity. It makes scrolling too slow.
-from now on, dropping is automatically enabled by having its listener. If set to null (or not set at all), it's turned off.

thijssteel commented:
Could you do the changing of the names to kotlin extensions and the changing of the code in separate commits? That would give a cleaner history

AndroidDeveloperLB commented:
@thijssteel I don't understand the request.
What is "changing of the names to kotlin extensions" ? 
One of the first things I've done is to convert to Kotlin. I think the only thing I've done before this (and committed) is to update the SDKs versions and remove the build tools version (which isn't needed anymore).
Here's the commit before conversion to Kotlin:
https://github.com/Quivr/Android-Week-View/pull/97/commits/472162d2e1a6c2d31d7584e31626e90f2b0502c7
And the one after :
https://github.com/Quivr/Android-Week-View/pull/97/commits/2b59062d24728f76d0979b84db1da36a8052ffe2

After that, everything is in Kotlin.

thijssteel commented:
Due to how git works, changing file names and file contents in one commit, it will be hard to follow the history of the code.
Reset the branch to the parent commit, but keep the changes and stash them.
Rename the .java files to .kt, and commit this. Then apply the stash and commit the other changes.
The diff of the second commit will also be easier to review.

AndroidDeveloperLB commented:
Well as I wrote, one of the first commits (actually it's the second) was a whole conversion to Kotlin. This means all Java files changed to Kotlin files.
So all of this is not needed. You can't compare old vs new, because everything has changed.

Plus, if you wish to see the differences, you can click on each of the commits I've done, and also see the history of my fork.

thijssteel commented:
Well, not everything has changed...
I believe that if the filename changes are done in a separate commit, git could recognise it.

AndroidDeveloperLB commented:
The most important files (the code itself) has changed completely (replaced by new files). Java isn't the same as Kotlin. Similar, but not the same. Plus the conversion has messed up the order in some cases, sadly.

It was a very delicate procedure to convert the code properly. For each bug I've found after that, I've checked if it existed before (and so far it always did).

Later I've cleaned the code and removed un-needed comments (for example of CTORs that don't exist anymore as they got merged).

AndroidDeveloperLB committed:
-added a way to put text on the left of the all-day events.
-Minor bugs fixes
-Minor optimizations to make drawing and scrolling smoother

AndroidDeveloperLB committed:
fixed background color of samples.

thijssteel commented:
https://github.com/Quivr/Android-Week-View/tree/kotlin-base

you could rebase on this branch, @hermansje fixed the commits

thijssteel commented:
just a suggestion

AndroidDeveloperLB commented:
@thijssteel I'm quite new to huge pull requests here.
I don't understand what you mean and want. I just offered my contribution. 
Please let me know when the pull request was accepted and is fully available here, so that I could delete my fork and use this one instead.

I think I will stop committing for now. It's quite ready for using in the app we develop.

Of course there are still things to improve here, such as better performance and support for RTL, but I was told to move on and start using what I've done so far.

thijssteel commented:
Okay, thanks for your contributions. See you in juli

AndroidDeveloperLB commented:
@thijssteel Thank you.
One thing I have forgotten though is to update the readme file, because I've modified some attributes and functions (added, deleted and renamed some).
The reason is that I've noticed some attributes don't do anything, and that some are confusing or just wrong.

AndroidDeveloperLB commented:
@thijssteel It seems I've caused a tiny bug that (all) the event rectangle isn't precisely positioned on its correct location. Will need to investigate it and fix it.
EDIT: fixed it and other issues I've found.

AndroidDeveloperLB committed:
fixed positioning of events and their lines.

AndroidDeveloperLB committed:
- fixed issue of not being able to add an all-day event with same start-time and end-time, because if the end time is the same, it should still be considered ok.
- removed deprecated long id, because it's too common for calendar services to use String instead

AndroidDeveloperLB committed:
Restored prefetching mechanism, fixing this issue:
https://github.com/Quivr/Android-Week-View/pull/86
https://github.com/Quivr/Android-Week-View/pull/89

AndroidDeveloperLB committed:
fixed possible OOM when setting a time of 0:00 to events, due to splitting them into multiple events

AndroidDeveloperLB committed:
fixed various issues with all-day events.
added CTOR for all-day event that takes a single day.

AndroidDeveloperLB committed:
moved implementation of textColorPicker to default one of WeekView.

AndroidDeveloperLB committed:
made WeekViewEvent to be extendable, yet made most of its fields final

AndroidDeveloperLB committed:
handled NPE in PrefetchingWeekViewLoader.

AndroidDeveloperLB committed:
added feature to set a default title for events that don't have a title, using `untitledEventText`

AndroidDeveloperLB committed:
allowed using extended classes of WeekViewEvent in various functions

AndroidDeveloperLB committed:
split the padding of the header into top+bottom
added getters to some of the inner fields, so that you can put views on top of the WeekView on the correct places.
replaced the time auto-update mechanism that used Handler to update every minute. Now uses proper BroadcastReceiver instead.

AndroidDeveloperLB committed:
optimized code a bit.
fixed issue of not getting correct lastVisibleDate.

AndroidDeveloperLB committed:
More optimizations, by caching formatted dates and times.

AndroidDeveloperLB committed:
fixed weird issue in sample about January events

AndroidDeveloperLB commented:
I've fixed a lot of issues and improved performance by about x2 of the library (mostly by using cached results). I'm sure there are more that can be optimized. 
I think I will have a break now to work on other things. 
Hope I will be able to add RTL support too, but this could take a long time.

AndroidDeveloperLB committed:
setting a default gap between overlapping events as 1dp, because when there are multiple of them from same calendar, it looks bad.
minor performance improvements.

AndroidDeveloperLB committed:
setting a default gap between overlapping events as 1dp, because when there are multiple of them from same calendar, it looks bad.
minor performance improvements.

AndroidDeveloperLB committed:
Merge branch 'develop' of https://github.com/AndroidDeveloperLB/Android-Week-View into develop

# Conflicts:
#	library/src/main/java/com/alamkanak/weekview/WeekView.kt

AndroidDeveloperLB committed:
added side subtitle text, so you can show the year in case it's not the current year.
minor improvements.

AndroidDeveloperLB committed:
targeting Android P

AndroidDeveloperLB committed:
trying to fix not being able to use the library.

AndroidDeveloperLB committed:
Updated SDKs.
Updated to use Android-X instead of support library.

AndroidDeveloperLB committed:
updated gradle version

AndroidDeveloperLB committed:
updated sdks

AndroidDeveloperLB committed:
Update README.md

nivritgupta commented:
@AndroidDeveloperLB  is your kotlin library has  drag and drop event features ? , I noticed someone trying to implement drag and drop features but its not actually working with existing events  , do you have any idea how we can achieve this feature ?

AndroidDeveloperLB commented:
@nivritgupta I didn't implement this feature, and when I converted (and before), I've noticed it made the UI very slow, so I think the implementation of it is very flawed and I don't recommend using this feature in its current form.
You are free to try, of course. It's available on the sample of my repository, just on the part that showed the original sample that I've forked from...

nivritgupta commented:
@AndroidDeveloperLB  did you know any alternative  calendar that  have the same functionality like android week view  with drag and drop events features ? I am trying to implement  drag and drop  features in "android week view" but I think its very hard to achieve this functionality with current  coding structure .

AndroidDeveloperLB commented:
@nivritgupta Sorry, I don't.
It was hard enough to even find a library to show the calendar nicely.
If you have a lot of time, I suggest considering making your own solution, because I think it might be hard to force something so complex to work the way you want. Plus, no matter how much I worked on this, I've always found somewhere that could be fixed or improved...

reilem commented:
Sorry for sitting on this PR for so long, but as you can probably understand a large PR like this is quite hard to review and we're afraid that the library might no longer behave as we expect it to. However, a kotlin conversion is definitely necessary for the future of this library and so I hope I find the time to test this. I've looked over everything quickly for now and I only have a few questions/concerns:

- I've seen the usage of double bang (!!) operators here and there. While I know these can work in certain cases I would like to avoid these as much as possible since it increases the risk of a null causing a crash. Even if you claim to have "checked it somewhere else", or that the exception gets "caught", these really damage the modifiability of the library. If someone accidentally removed these checks it can cause unnecessary side effects.

- Does this PR also fix API 28 compatibility like the PR #119 does?

AndroidDeveloperLB commented:
@reilem Ok I have comments to the comments:

- The conversion to Kotlin itself has replaced all Java files, so it makes sense that a lot has changed. Everything has changed. 
- I didn't only convert to Kotlin. I've also fixed a lot of bugs and improved performance. 
- The `!!` is used like on Java, so there is no need to worry about NPE more than what you worry about it on Java. In fact, because I converted to Kotlin, I had to go over many places to make sure the conversion is correct, because nullability is one of the main issues when converting to Kotlin.
- About the fix of API 28, already fixed. In fact, I'm the one who asked about what's the best way to fix it, [here ](https://stackoverflow.com/q/50231950/878126). Was much long before the post you've shown ... :)

I think you should embrace this change. Sooner or later it's nicer to have it all in Kotlin. I think what i did made the library more mature. 

I've also written where it can be better, what else should be fixed...

reilem commented:
Alright thanks, currently we are working quite heavily on another project but if I have time I will try to locally merge your PR and test all important functionality.

As for the double bang operator, my concern is exactly what you stated. Kotlin adds nice functionality and lots of improvements, and we shouldn't rely on old java logic for null checking. Since all you need to do is add a `!= null` check and it will be automatically casted to a non-optional. I will probably make this change either way so don't worry if you don't have time to do it.

AndroidDeveloperLB commented:
@reilem It's not that you should avoid "!!". It's just that it means that you are aware of what you are doing, that it shouldn't be null when you reach there. 
Of course things can get wrong, but the same things could be wrong even if you add null check everywhere. It really depends on the case.
This is especially true for custom views, where almost everything is on a single thread anyway (the UI thread).
What do you mean by "I will probably make this change either way so don't worry if you don't have time to do it." ?

reilem commented:
All I really want to say is that adding a `!= null` check is simple, easy and takes literally 2 seconds and avoids a many great deal of issues in the future. It allows for graceful degradation instead of simply crashing the app and should be used whenever possible. If needed you can throw an exception when it is null so that developers are still able to detect the issue.

I have looked at this situation and understand that right now it seems obvious to you, but to future contributors this sort of coding behaviour can be very confusing. I have been in a few situations where a new programmer joins a team and causes these types of bugs due to previous developer negligence.

And what I mean is that since I am very anal about these type of safety measures haha. And that there's a good chance I will be adding the `!= null` check because it is simply my coding philosophy.

Edit: I’ve removed some of our comments below so that this thread stays on-topic.

AndroidDeveloperLB commented:
@reilem It really depends on the case. Adding null checks everywhere doesn't make sense in all cases. 
And sorry, but I don't know the term " I am very anal" .

reilem commented:
Yes I understand. Oh and sorry, "I am anal about something" means that I am a bit "obsessive" about it haha. 😄

AndroidDeveloperLB commented:
@reilem I've really never heard this term. English is weird. I was sure you wrote a typo because it doesn't make sense when I think about the verbal translation...

reilem commented:
This is the official explanation :) https://en.wikipedia.org/wiki/Anal_retentiveness

AndroidDeveloperLB commented:
@reilem Seems disgusting .
